File: tobas_setup_assistant/src/tobas_setup_assistant/parameter_getters/int_range.py
from typing import override
from typing import Tuple, Optional
from PyQt5.QtCore import pyqtSignal, pyqtSlot
from PyQt5.QtWidgets import QHBoxLayout

from .base import ParamGetterWidget
from .utils import IntGetter


class ParamGetterWidget_IntRange(ParamGetterWidget[Tuple[int, int]]):
    value_changed = pyqtSignal(int, int)

    def __init__(
        self,
        param_name: str,
        description_text: Optional[str] = None,
        minimum: int = -(10**9),
        maximum: int = +(10**9),
        single_step: int = 1,
        default: Tuple[int, int] = (0, 0),
        suffix: str = "",
    ) -> None:
        super().__init__(param_name, description_text)

        cols = QHBoxLayout()
        self._rows.addLayout(cols)

        self._min = IntGetter("min", minimum, maximum, single_step, default[0], suffix)
        cols.addWidget(self._min)

        self._max = IntGetter("max", minimum, maximum, single_step, default[1], suffix)
        cols.addWidget(self._max)

        self._min.value_changed.connect(self._on_value_changed)
        self._max.value_changed.connect(self._on_value_changed)

    def min(self) -> int:
        return self._min.get()

    def max(self) -> int:
        return self._max.get()

    @override
    def get(self) -> Tuple[int, int]:
        return self.min(), self.max()

    @override
    def set(self, src: Tuple[int, int]) -> None:
        self._min.set(src[0])
        self._max.set(src[1])

    def is_valid(self) -> bool:
        return self.min() <= self.max()

    @pyqtSlot(int)
    def _on_value_changed(self, value: int) -> None:
        self.value_changed.emit(self.min(), self.max())

File: tobas_gui/tobas_setup_assistant/src/param_getters/int_range.cpp
#include "tobas_setup_assistant/param_getters/int_range.hpp"

namespace gui
{
namespace setup_assistant
{
ParamGetterWidget_IntRange::ParamGetterWidget_IntRange(const QString& param_name, const QString& description_text)
  : super(param_name, description_text)
{
  auto cols = new QHBoxLayout();
  rows_->addLayout(cols);

  min_ = new IntGetter("min");
  cols->addWidget(min_);

  max_ = new IntGetter("max");
  cols->addWidget(max_);

  connect(min_, &IntGetter::valueChanged, this, &self::onValueChanged);
  connect(max_, &IntGetter::valueChanged, this, &self::onValueChanged);
}

std::pair<int, int> ParamGetterWidget_IntRange::getValue() const
{
  return { min(), max() };
}

bool ParamGetterWidget_IntRange::setValue(const std::pair<int, int>& src)
{
  min_->setValue(src.first);
  max_->setValue(src.second);
  return true;
}

void ParamGetterWidget_IntRange::setMinimum(int minimum)
{
  min_->setMinimum(minimum);
  max_->setMinimum(minimum);
}

void ParamGetterWidget_IntRange::setMaximum(int maximum)
{
  min_->setMaximum(maximum);
  max_->setMaximum(maximum);
}

void ParamGetterWidget_IntRange::setSingleStep(int single_step)
{
  min_->setSingleStep(single_step);
  max_->setSingleStep(single_step);
}

void ParamGetterWidget_IntRange::setSuffix(const QString& suffix)
{
  min_->setSuffix(suffix);
  max_->setSuffix(suffix);
}

int ParamGetterWidget_IntRange::min() const
{
  return min_->getValue();
}

int ParamGetterWidget_IntRange::max() const
{
  return max_->getValue();
}

bool ParamGetterWidget_IntRange::isValid() const
{
  return min() <= max();
}

void ParamGetterWidget_IntRange::onValueChanged(int)
{
  Q_EMIT valueChanged({ min(), max() });
}
}  // namespace setup_assistant
}  // namespace gui
